File: sdk/go/pkg/artifact/bun.go
package artifact

import (
	"fmt"

	api "github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/api/artifact"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/config"
)

const defaultBunVersion = "1.3.10"

func Bun(context *config.ConfigContext) (*string, error) {
	name := "bun"

	system := context.GetTargetStr()

	var sourceTarget string
	switch system {
	case "aarch64-darwin":
		sourceTarget = "darwin-aarch64"
	case "aarch64-linux":
		sourceTarget = "linux-aarch64"
	case "x86_64-darwin":
		sourceTarget = "darwin-x64"
	case "x86_64-linux":
		sourceTarget = "linux-x64-baseline"
	default:
		return nil, fmt.Errorf("unsupported %s system: %s", name, system)
	}

	sourceVersion := defaultBunVersion
	sourcePath := fmt.Sprintf("https://sdk.vorpal.build/source/bun-%s-%s.zip", sourceVersion, sourceTarget)

	source := NewArtifactSource(name, sourcePath).Build()

	stepScript := fmt.Sprintf(`mkdir -p "$VORPAL_OUTPUT/bin"
cp -p "./source/%s/bun-%s/bun" "$VORPAL_OUTPUT/bin/bun"
chmod +x "$VORPAL_OUTPUT/bin/bun"
`, name, sourceTarget)

	step, err := Shell(context, []*string{}, []string{}, stepScript, nil)
	if err != nil {
		return nil, err
	}

	return NewArtifact(name, []*api.ArtifactStep{step}, config.SYSTEMS).
		WithAliases([]string{fmt.Sprintf("%s:%s", name, sourceVersion)}).
		WithSources([]*api.ArtifactSource{&source}).
		Build(context)
}

File: sdk/python/src/vorpal_sdk/artifact/bun.py
"""Bun runtime artifact (prebuilt binary from a zip archive).

Mirrors ``sdk/typescript/src/artifact/bun.js``.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from vorpal_sdk.artifact import Artifact, ArtifactSource
from vorpal_sdk.step import shell
from vorpal_sdk.system import SYSTEMS, get_system_str

if TYPE_CHECKING:
    from vorpal_sdk.context import ConfigContext

DEFAULT_BUN_VERSION = "1.3.10"


class Bun:
    """Builder for the Bun runtime artifact."""

    def __init__(self) -> None:
        self._version = DEFAULT_BUN_VERSION

    def with_version(self, version: str) -> Bun:
        self._version = version
        return self

    def build(self, context: ConfigContext) -> str:
        name = "bun"
        system = context.get_system()
        system_str = get_system_str(system)

        if system_str == "aarch64-darwin":
            source_target = "darwin-aarch64"
        elif system_str == "aarch64-linux":
            source_target = "linux-aarch64"
        elif system_str == "x86_64-darwin":
            source_target = "darwin-x64"
        elif system_str == "x86_64-linux":
            source_target = "linux-x64-baseline"
        else:
            raise ValueError(f"unsupported {name} system: {system}")

        source_version = self._version
        source_path = (
            f"https://sdk.vorpal.build/source/"
            f"bun-{source_version}-{source_target}.zip"
        )

        source = ArtifactSource(name, source_path).build()

        step_script = f"""mkdir -p "$VORPAL_OUTPUT/bin"
cp -p "./source/{name}/bun-{source_target}/bun" "$VORPAL_OUTPUT/bin/bun"
chmod +x "$VORPAL_OUTPUT/bin/bun"
"""
        steps = [shell(context, [], [], step_script, [])]

        return (
            Artifact(name, steps, SYSTEMS)
            .with_aliases([f"{name}:{source_version}"])
            .with_sources([source])
            .build(context)
        )
